numpy/core/src/multiarray/calculation.c

```diff
@@ -266,6 +319,221 @@ PyArray_ArgMin(PyArrayObject *op, int axis, PyArrayObject *out)
+            * When axis is `None` in the input then all dimensions
+            * of out array should be 1. Note that for comparing number
+            * of dimensions we have used `op_NDIM` because it is the actual
+            * number of dimensions in the input. `ap` looses one dimension
+            * due to previous code.
+            */
+            shape_check = PyArray_NDIM(out) == op_NDIM;
+            for( int dim = 0; shape_check && dim < PyArray_NDIM(out); dim++ ) {
+                shape_check = PyArray_DIMS(out)[dim] == 1;
+            }
+        } else {
+            /*
+            * Standard case, the out array dimensions should be same
+            * as input array except the last one which should be 1.
+            */
+            shape_check = (PyArray_NDIM(out) == PyArray_NDIM(ap)) &&
+                           PyArray_CompareLists(PyArray_DIMS(out), PyArray_DIMS(ap),
+                                                 PyArray_NDIM(out) - 1) && 
+                           (PyArray_DIMS(out)[PyArray_NDIM(out) - 1] == 1);
+        }
+        if (!shape_check) {
+            PyErr_SetString(PyExc_ValueError,
+                    "output array does not match result of np.argmin.");
+            goto fail;
+        }
+        rp = (PyArrayObject *)PyArray_FromArray(out,
+                              PyArray_DescrFromType(NPY_INTP),
+                              NPY_ARRAY_CARRAY | NPY_ARRAY_WRITEBACKIFCOPY);
+        if (rp == NULL) {
+            goto fail;
+        }
+    }
+
+    NPY_BEGIN_THREADS_DESCR(PyArray_DESCR(ap));
+    n = PyArray_SIZE(ap)/m;
+    rptr = (npy_intp *)PyArray_DATA(rp);
+    for (ip = PyArray_DATA(ap), i = 0; i < n; i++, ip += elsize*m) {
+        arg_func(ip, m, rptr, ap);
+        rptr += 1;
+    }
+    NPY_END_THREADS_DESCR(PyArray_DESCR(ap));
+
+    Py_DECREF(ap);
+    /* Trigger the UPDATEIFCOPY/WRITEBACKIFCOPY if necessary */
+    if (out != NULL && out != rp) {
+        PyArray_ResolveWritebackIfCopy(rp);
+        Py_DECREF(rp);
+        rp = out;
+        Py_INCREF(rp);
+    }
+
+    /*
+    * The following code takes the axis back to its original
+    * position which was placed in the last dimension earlier.
+    * This can be interpreted as reversing the previously taken
+    * transpose of output, which was taken to travel through
+    * the desired axis by just increasing the pointer by 1.
+    */
+    npy_intp dims[NPY_MAXDIMS];
+    newaxes.ptr = dims;
+    newaxes.len = PyArray_NDIM(rp);
+    npy_intp k; 
+    for (i = 0, k = 0; i < PyArray_NDIM(rp); i++ ) {
+        if( i == axis ) {
+            continue;
+        }
+        dims[i] = k;
+        k++;
+    }
+    dims[axis] = PyArray_NDIM(rp) - 1;
+    op = (PyArrayObject *)PyArray_Transpose(rp, &newaxes);
+    if (op == NULL) {
+        Py_DECREF(rp);
+        return NULL;
+    }
+    Py_DECREF(rp);
+    rp = op;
+    Py_INCREF(rp);
+    return (PyObject *)rp;
+
+ fail:
+    Py_DECREF(ap);
+    Py_XDECREF(rp);
+    return NULL;
+}
+
```

Comment:
This approach takes permutes the dimensions of the array twice, first for traversing the desired axis and then for aligning the dimensions with that of the input. I believe, permutation (`PyArray_Transpose`) is linear operation in the number of dimensions (O(ndim)).